### ISSUE TITLE
print: Fix printing for several cases

### ISSUE THREAD

**Issue by georgebisbas** [xdslproject/xdsl#275]
Currently, Printer().print_op is broken for a few cases, example:

```bash

from xdsl.ir import OpResult, SSAValue
from xdsl.dialects.builtin import i32, IntegerType, IndexType
from xdsl.dialects.memref import (Alloc, Alloca, Dealloc, Dealloca, MemRefType,
                                  Load, Store)

def test_memref_store_i32_with_dimensions():
    i32_memref_type = MemRefType.from_type_and_list(i32, [2, 3])
    memref_ssa_value = SSAValue(i32_memref_type)
    i32_ssa_value = SSAValue(IntegerType.from_width(32))
    index1 = SSAValue(IndexType)
    index2 = SSAValue(IndexType)
    store = Store.get(i32_ssa_value, memref_ssa_value, [index1, index2])
    
    from xdsl.printer import Printer
    Printer().print_op(store)
```

```bash
memref.store(*** TypeError: unhashable type: 'SSAValue'
```

**Comment by math-fehr** [comment]
For attributes, we can fix this easily.
For operations though, I guess the first step would be to print operations in their generic form, and just write something like `%<FREE>` for each operand appearing outside the scope.
For printing them without the regions, I'm not sure yet how to do that (at least not without printing the custom format).
I'll write a simple PR for the first cases at least!

**Comment by webmiche** [comment]
I am a bit confused why this fails. I would expect this to print `%UNKNOWN` and the `did you forget to add an operation`-error. At least I think that these are the cases where we initially wanted to print this error.

**Comment by webmiche** [comment]
Ah nvm, I see it. Usually, `SSAValue`s are either `BlockArg`s or `OpResult`s, both of which are hashable. `SSAValue` itself on the other hand is not hashable, but the printer relies on the hash in order to look it up in the dictionary of encountered `SSAValue`s in order to get its value.

IMO, circumventing this particular error will mean (a) constructing `OpResult`s instead of `SSAValue`s or, more fundamentally, (b) make SSAValue hashable or cahnge the printer to (b1) catch the exception or (b2) restructure to not rely on the `SSAValue`s to be hashable.

@math-fehr do you have an idea why `SSAValue` should not be hashable? Do we just not want the class instantiated in the IR (I guess inheriting from `ABC` somewhat entails that)?

**Comment by math-fehr** [comment]
Ah, I completely missed the point that George wanted to instantiate directly these variables.
I think we have two options here:
* Either we define a new `TestConstant` operation that can instantiate anything. Then, we define a new printer that can print a single operation in isolation (by printing `%UNKNOWN`, without an error message).
* Or we allow `SSAValue` to be instantiated like George did, which I think is dangerous because `SSAValue` should not be used like this, and then we define the same printer with `%UNKNOWN`.

To be honest, I think there is another question here. Why do we want to test using the custom format? If we are writing things in Python, I would suggest to test it as well with Python, not with the custom format.
If we want to test the custom format, I would test it in `FileCheck` rather than by printing.

However, for debugging purposes, I still thing it's a good idea to print a single operation easily.

**Comment by webmiche** [comment]
I think this is the way I constructed things for the memref tests as well, i.e. using SSAValue directly, because I wasn't thinking of this problem, and I still feel that it makes sense for the tests where I used it. In some sense, it is a testing substitute to check invariants on the object under test.

I would prefer the first decision here if we want to allow this type of behavior. 

I agree with the point on printing. I also feel that there are a lot of pytest tests that we should rewrite s.t. they do not rely on printing stuff in order to check invariants.
